### ISSUE TITLE
CRITs samples API only supports md5 hash type on metadata post

### ISSUE THREAD

**Issue by treyka** [crits/crits#362]
I'm trying to import samples via the API but it only accepts md5 hashes.

**Comment by mgoffin** [comment]
Yes. Currently MD5 is a required field for all Samples (it is not trivial to change this). All other hashes are calculated off the binary.

Someone in the community has made changes to allow for additional hashes to be included along with the MD5 on upload. They are still working with it and aren't entirely sure they can share the code yet. It's still going through the works so I wouldn't count on it any time soon.

**Comment by thelok** [comment]
Don't know if you were referring to me but I was looking at allowing for other hashes such as SHA hashes... but I think because of the ties to Mongo records and the fact that the ObjectIDs are MD5s and how the chunk sample data was stored, it looked like it was going to be a HUGE pain. I gave up pursuing this.

**Comment by mgoffin** [comment]
Not referring to you, but yes this is a pain. To clarify some of the stuff you mentioned:

* MD5 is the *shard key* for Samples, PCAPs, Certificates, etc. in a cluster, which cannot be changed for a given database without wiping all of the data (or moving it to another collection manually which has a different shard key)
* GridFS determines duplicates based on MD5 (not our code, this is in Mongo)

To make this happen we need to do the following:

* Move away from using values other than ObjectId for accessing content, specifically through URL parameters (details pages for Samples use the MD5 whereas Indicators use the ObjectId which is preferable).
* Move away from using calculated hashes as shard keys and move over to using ObjectId Hashes as a shard key (http://docs.mongodb.org/manual/tutorial/create-a-hashed-index/)
  * In the past when we moved the MD5 of a Sample to be a top-level field, we had to create a new collection (sample), set the new shard key on it, and then manually write every sample over to that collection (this was the sample migration script for 3.0...https://github.com/crits/crits_services/blob/master/crits_scripts/scripts/samples_3_0.py)
  * Steps would be:
    * Create new collections (we can go back to "samples", then we would need to use pcap, certificate, object for the rest of the ones that are sharded by MD5)
    * Set the hashed ObjectId as the shard key on those collections.
    * Write scripts to migrate the data to the new collections.
    * Modify the entire code base to now leverage ObjectId for accessing items and make novelty URLs for things like using a hash to find a sample:
      * /samples/details/md5/themd5/
      * /samples/details/sha1/thesha1/
      * etc.

A lot of work :( That's why the push from this community member was to just expose adding other hashes on upload but still require the MD5. Much easier, but still limiting for those who do not have the MD5.

**Comment by Lambdanaut** [comment]
Has there been any progress on this issue over the last year?

**Comment by WesQuinn** [comment]
I am interested in hearing more about the status of it too.

Thanks,
Wes Quinn

**Comment by mokarimi** [comment]
Any progress on this?  The data that we get is limited to SHA256 but Sample still requires MD5.

**Comment by mgoffin** [comment]
No there's been no updates. Given the effort involved in the comment I made above it hasn't been done at this time. If anyone wants to do the work and contribute it back, we'd be happy to review.
